### PR TITLE
Test ini file common operation and Update parsecfg.nim 

### PR DESCRIPTION
**Pull request by lihf8515**

### CODE REVIEW

**Comment by Araq** [comment]
@dom96 looks fine to me but stdlib is yours.

File: lib/pure/parsecfg.nim [hunk 1]
```diff
@@ -22,10 +22,66 @@
 ##
 ## .. code-block:: nim
 ##     :file: examples/parsecfgex.nim
-
+## 
+## Examples
+## --------
+##
+## This is an example of how a configuration file
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: "how a" -> "of a"

File: lib/pure/parsecfg.nim [hunk 2]
```diff
@@ -22,10 +22,66 @@
...
+## .. include:: config.ini
+## 
+##     [Package]
+##     name = "hello"
+##     [Author]
+##     name = "lihf8515"
+##     qq = "10214028"
+##     email = "lihaifeng@wxm.com"
+## 
+## Create configuration file
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: "How to create a configuration file"

File: lib/pure/parsecfg.nim [hunk 3]
```diff
@@ -22,10 +22,66 @@
...
+##     import parsecfg
+##     var dict=newCfg()
+##     dict.setSectionKey("Package","name","hello")
+##     dict.setSectionKey("Author","name","lihf8515")
+##     dict.setSectionKey("Author","qq","10214028")
+##     dict.setSectionKey("Author","email","lihaifeng@wxm.com")
+##     dict.writeCfgFile("config.ini")
+## 
+## Read configuration file
+## A number of the same name keys or section, will take the value of last
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: Move this below the `=========`.

Change to "Reading a configuration file"

File: lib/pure/parsecfg.nim [hunk 4]
```diff
@@ -22,10 +22,66 @@
...
+##
+##     import parsecfg
+##     var dict = loadCfgFile("config.ini")
+##     var pname = dict.getSectionKey("Package","name")
+##     var name = dict.getSectionKey("Author","name")
+##     var qq = dict.getSectionKey("Author","qq")
+##     var email = dict.getSectionKey("Author","email")
+##     echo pname & "\n" & name & "\n" & qq & "\n" & email
+## 
+## Modify configuration file
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: "Modifying a configuration file"

File: lib/pure/parsecfg.nim [hunk 5]
```diff
@@ -22,10 +22,66 @@
...
+## =========================
+## .. code-block:: nim
+## 
+##     import parsecfg
+##     var dict = loadCfgFile("config.ini")
+##     dict.setSectionKey("Author","name","lhf")
+##     dict.writeCfgFile("config.ini")
+##
+## Delete the key in the configuration file
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: "Deleting a section key in a configuration file"

File: lib/pure/parsecfg.nim [hunk 6]
```diff
@@ -288,19 +344,19 @@ proc rawGetTok(c: var CfgParser, tok: var Token) =
   else: getSymbol(c, tok)
 
 proc errorStr*(c: CfgParser, msg: string): string {.rtl, extern: "npc$1".} =
-  ## returns a properly formatted error message containing current line and
+  ## returns a properly formated error message containing current line and
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: The old spelling is correct.

File: lib/pure/parsecfg.nim [hunk 7]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
     result.kind = cfgError
     result.msg = errorStr(c, "invalid token: " & c.tok.literal)
     rawGetTok(c, c.tok)
+    rawGetTok(c, c.tok)
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: Why is this now done twice?

File: lib/pure/parsecfg.nim [hunk 8]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
     result.kind = cfgError
     result.msg = errorStr(c, "invalid token: " & c.tok.literal)
     rawGetTok(c, c.tok)
+    rawGetTok(c, c.tok) 
+
+# --------------------- Configuration file related operations ---------------------
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: Word wrap at 80 chars please.

File: lib/pure/parsecfg.nim [hunk 9]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+
+proc newCfg*(): OrderedTableRef[string, OrderedTableRef[string, string]] =
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: Please define the type `OrderedTableRef[string, OrderedTableRef[string, string]]` above, name it "Config" or something:

```nim
type
  Config = OrderedTableRef[string, OrderedTableRef[string, string]]
```

Comment (dom96) [inline lib/pure/parsecfg.nim]: Also rename this procedure to ``newConfig``.

File: lib/pure/parsecfg.nim [hunk 10]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+  ## Create Cfg table.
+  ## When you create a new configuration file, you need to use it first.
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: Change to:

> Create a new configuration table. Useful when wanting to create a configuration file.

File: lib/pure/parsecfg.nim [hunk 11]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+  result = newOrderedTable[string, OrderedTableRef[string, string]]()
+
+proc loadCfgFile*(filename: string): OrderedTableRef[string, OrderedTableRef[string, string]] =
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: rename to `loadConfig`.

File: lib/pure/parsecfg.nim [hunk 12]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+  ## Load the specified configuration file into table.
+  ## When you need to read, modify, and delete operations, you need to use it first.
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: Change to:

> Load the specified configuration file into a new `Config` instance.

File: lib/pure/parsecfg.nim [hunk 13]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+    elif s[i] == '\c':
+      d.add(r"\n")
+    elif s[i] == '\L':
+      d.add(r"\n")
+    else:
+      d.add(s[i])
+    inc(i)
+  result = d
+
+proc writeCfgFile*(dict: OrderedTableRef[string, OrderedTableRef[string, string]], filename: string) =
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: Rename to `writeConfig`.

File: lib/pure/parsecfg.nim [hunk 14]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+  ## Writes the contents of the table to the specified configuration file.
+  ## Note: write to ignore the comment
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: What does "Note: write to ignore the comment" mean?

File: lib/pure/parsecfg.nim [hunk 15]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+  var file: File
+  if file.open(filename, fmWrite):
+    try:
+      var section, key, value, kv, segmentChar:string
+      for pair in dict.pairs():
+        section = pair[0]
+        if section != "": ## Not general section
+          if not allCharsInSet(section, SymChars): ## Non system character
+            file.writeLine("[\"" & section & "\"]")
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: Can the parser parse `["blah"]` properly?

File: lib/pure/parsecfg.nim [hunk 16]
```diff
@@ -359,3 +415,139 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+            else:
+              kv.add(segmentChar)
+              kv.add(value)
+          file.writeLine(kv)
+    except:
+      raise
+    finally:
+      file.close()
+
+proc getSectionKey*(dict: OrderedTableRef[string, OrderedTableRef[string, string]], section, key: string): string =
```
Comment (dom96) [inline lib/pure/parsecfg.nim]: I think this should be called `getSectionValue`, no?

**Comment by dom96** [comment]
It's a nice addition but it definitely needs some polish.

**Pull request closed without merging**